File: android/app/src/main/kotlin/com/example/hotel_reception/procesor/Processor.kt
package com.example.hotel_reception.procesor

import android.util.Log
import com.example.hotel_reception.procesor.convertors.convertToUpdate
import com.example.hotel_reception.procesor.model.*
import java.util.*

class Processor(
    private val world: World,
    private val notifier: ChannelNotifier,
    initialData: InitialData,
) :
    WorldDateChanged {
    private val hotel: Hotel

    init {
        world.subscribeOnDateChanges(this)
        hotel = Hotel.buildHotel(initialData.service, initialData.roomByType, world)
        world.startWorld()
    }

    override fun onDateChanged() {
        hotel.checkDesk()
        hotel.checkInPersons(PersonRandomizer.generatePersonQuery(world.getDate()))
        if (world.getDayOfMonth() == 30) hotel.payTaxes()
        Log.i(logTag, "date: ${world.getDate()} balance: ${hotel.balance}")
        notifier.notifyChannel(hotel.convertToUpdate())
    }
}


interface ChannelNotifier {
    fun notifyChannel(data: String)
}

File: ios/Runner/Processor/Processor.swift
//
//  Processor.swift
//  Runner
//
//  Created by Serhii Rakovskyi on 18.08.2022.
//

import Foundation
class Processor : WorldChangedListener {
    private let world: World
    private let notifier: ChannelNotifier
    private let hotel:Hotel
    
    init(world: World, notifier: ChannelNotifier, initData:InitialData) {
        self.world = world
        self.notifier = notifier
        self.hotel = Hotel.build(service: initData.service, roomByType: initData.roomByType, worldsClock: world)
        world.subscribeOnDateChanges(listener: self)
        print("starting world")
        world.startWorld()
    }
    
    func onDateChanged() {
        print("onDateChanged")
        hotel.checkDesk()
        hotel.checkInPersons(persons: PersonRandomizer().generatePersonQuery(date: hotel.getToday()))
        if  world.getDayOfMonth() == 30 {
                        hotel.payTaxes()
        }
        notifier.notifiChannel(data: HotelConverter.convertToUpdate(hotel: hotel))
    }
    
}


protocol ChannelNotifier {
    func notifiChannel(data:String)
}
